rustup-init: Support updating the installed toolchain

kinnison opened the pull request:
If the user specifies any of --default-toolchain (not none)
or --target or --component when invoking `rustup-init` then
we will update the existing installation toolchains, rather
than the previous behaviour which was to skip toolchain
installation.  This is nominally a breaking change but likely
ends up adhering closer to the principle of least surprise.

At this point, the profile will not be applied, so a change
of profile from minimal to default will not cause additional
components to end up being installed.

In theory this fixes #2194 but will need checking with @flaviojs
Also @lzutao filed another bug, this closes #2215 too.

kinnison committed:
rustup-init: Support updating the installed toolchain

If the user specifies any of --default-toolchain (not none)
or --target or --component when invoking `rustup-init` then
we will update the existing installation toolchains, rather
than the previous behaviour which was to skip toolchain
installation.  This is nominally a breaking change but likely
ends up adhering closer to the principle of least surprise.

At this point, the profile will not be applied, so a change
of profile from minimal to default will not cause additional
components to end up being installed.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

tesuji commented:
I wonder whether we could integrate parsing toolchains string to a type.
And use that type instead of processing raw string over and over again
in whole crate.

kinnison commented:
@lzutao We do have a ToolchainDesc in `dist.rs` and I suppose we could begin to rework APIs to expect one of those but it pushes the handling of bad toolchain names outward which might mean we end up with more spots needing to deal with that specific error rather than more general handling in the core.

rbtcollins commented:
hmm, so before I even check the code I will note that this will likely result in performance regression bugs on windows until we liaise with Microsoft's Defender team again to get rustup-init included in their scan-mode heuristics as well (I don't believe it currently is - I may be wrong).

rbtcollins commented:
Hmm, rustup-init already knows how to install toolchains, we must have had it included in the heuristics already. Nevermind - or at least, lets treat that as a separate issue. Looking at the code now.

flaviojs commented:
Installing/updating according to the arguments is what I originally expected, so this fixes #2188 (my initial issue).

From the changes it seems I'll be able to know if the toolchain was installed or not from the messages, so this fixes #2194.

rbtcollins commented on src/cli/self_update.rs:
```diff
@@ -746,21 +748,29 @@ fn maybe_install_rust(
+    // if they specify something directly, or we have no default, then we install
+    // a toolchain (updating if it's already present) and then if neither of
+    // those are true, we have a user who doesn't mind, and already has an
+    // install, so we leave their setup alone.
```
I don't think this is quite right. What if they ask for no tool chain and a component? Should that silently do nothing or error?

rbtcollins replied:
Well, it won't be silent, but it won't process the invalid input either.

rbtcollins replied:
other than this, the PR looks good

kinnison replied:
I think that's a good point.  I'd say it's part of `setup_mode.rs` to verify that though.  I've made a note to try and write a test into there.

kinnison replied:
I take that back, clearly it belongs after the user might have made a change.  If they specify targets/components on the CLI and then set the toolchain to none in the interactive prompts, then it still ought to warn.

kinnison replied:
I've pushed a small change, how does that feel to you?

kinnison committed:
rustup-init: Warn user if -c or -t is ignored

If the user specifies the default toolchain of 'none' but also
has specified components or targets then they will be ignored.
As such, warn if this is the case so that people can diagnose
things if their expectations fail.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison commented:
@rbtcollins Are you okay with the current state of things?  I'd like to get this merged if so.

rbtcollins commented on src/cli/self_update.rs:
```diff
@@ -746,21 +748,43 @@ fn maybe_install_rust(
+        let toolchain_str = toolchain.unwrap_or("stable");
         let toolchain = cfg.get_toolchain(toolchain_str, false)?;
+        if toolchain.exists() {
+            warn!("Updating existing toolchain, profile choice will be ignored");
```
Is this missing logic, or not well defined?

kinnison replied:
At the moment it's not clear what it'd mean - should it only *add* missing components?  Should it make it *be* only the components for the profile (e.g. if you say `minimal` should it remove anything else?) - as such, this is punted just like it is if you do `rustup toolchain install ...` after it's already installed, the profile will be ignored.

flaviojs replied:
`rustup set` says "The default components installed" about profile.
`rustup component` says we can add and remove components

Based on that I would assume that the profile only determines what is added on install. If the toolchain already exists I would expect it to add missing components and leave the rest alone.

Since the rust-init option is called `--profile` (no default) I would also assume it only affects that particular call. No idea if that is the current behavior.

kinnison replied:
So your proposal would be "If the user explicitly invoked `rustup toolchain install` (or `rustup-init` as is considered in this PR) then any components missing from the profile of an already installed toolchain should explicitly be added in.  I *think* that could be consistent but wouldn't be behaviour belonging in *this* PR.  I'd be up for an issue about that being opened though, if either of you agree with my restating of things.

rbtcollins replied:
I'm happy with 'this is not well defined yet, lets open a new issue to discuss'.

That proposed behaviour might be it, but another one might be to coerce back to the profiles exact definition - lets work through it elsewhere.
